---
title: "Pseudobulk analysis of crypts"
author: "Ricardo Ramirez"
date: "2022-12-15"
output: 
  html_document:
      code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
library(tidyverse)
library(scuttle)
library(edgeR)
library(Seurat)
library(knitr)
library(rmarkdown)
```

## Experimental design

We have been provided with groups of spots that belong to a certain crypt within a visium slide (n = 2). Within each slide there are two types of crypts (healthy and disease) and each slide represents a different condition.

The objective is to:

1) Identify the differentially expressed genes between types of crypts
2) Assess if there is an interaction with the different conditions (not detached from technical since there are no replicates of diets)

### Extract count matrix and meta data of original data

```{r}
d6_data <- readRDS("./R_objects/se.helminth.d6")
# Count data
DefaultAssay(d6_data) <- "RNA"
GEX_mat <- Seurat::GetAssayData(d6_data,assay = "RNA", slot = "counts")
GEX_mat <- GEX_mat[grepl("refdata-gex-mm10",rownames(GEX_mat)),]
rownames(GEX_mat) <- gsub("refdata-gex-mm10----------", "", rownames(GEX_mat))

# Meta data
meta_data <- d6_data@meta.data %>%
  rownames_to_column("raw_spot_id") %>%
  dplyr::mutate(spot_id = strsplit(raw_spot_id,
                                   "_") %>%
                  map_chr(., ~.x[[1]])) %>%
  dplyr::select(raw_spot_id, spot_id, diet, day, section_id)
```

### Spots of each crypt were provided

```{r, message=FALSE, warning= FALSE}
list_spots <- map(list.files("./ann/",full.names = T) %>% set_names,
    read_csv) %>%
  enframe() %>%
  unnest()

colnames(list_spots) <- c("file","spot_id", "ctrl_crypt", "granuloma_crypt")

list_spots <- list_spots %>% 
  dplyr::filter(!is.na(ctrl_crypt) | !is.na(granuloma_crypt)) %>%
  pivot_longer(-c(file,spot_id),names_to = "crypt_type", values_to = "id") %>%
  na.omit() %>%
  dplyr::mutate(diet = ifelse(grepl("STD", file), "STD", "GW")) %>%
  dplyr::select(-file)

meta_data <- meta_data %>%
  left_join(list_spots, by = c("spot_id", "diet")) %>%
  na.omit() %>%
  arrange(crypt_type, id) %>%
  dplyr::mutate(id = paste0(diet, "_", id))

# Check for duplicated spots during manual annotation

duplicated_spots <- meta_data$raw_spot_id[duplicated(meta_data$raw_spot_id)]

meta_data <- meta_data %>%
  dplyr::filter(!raw_spot_id %in% duplicated_spots)

coldata <- meta_data %>%
  dplyr::select(id, diet, crypt_type) %>%
  unique()

write_csv(meta_data, "./pub_results/spot_annotation.csv")
```

### Create pseudobulk profiles

```{r}
GEX_mat <- GEX_mat[,meta_data$raw_spot_id]

pb_GEX_mat <- scuttle::sumCountsAcrossCells(x = GEX_mat,
                                            ids  = meta_data %>% 
                                              dplyr::select(raw_spot_id, id) %>% 
                                              column_to_rownames("raw_spot_id") %>% 
                                              DataFrame())

pb_GEX_mat@colData <- pb_GEX_mat@colData %>%
  as.data.frame() %>%
  left_join(coldata, by = "id") %>%
  DataFrame()

assay(pb_GEX_mat, "counts") <- assay(pb_GEX_mat, "sum")
```

### Filter genes based on expression

Here we are strict in the sense that we have 6 pseudobulk profiles per diet and granuloma type, hence we will keep genes that have at least 1 read in 6 samples of a given group

```{r}
keep <- edgeR::filterByExpr(pb_GEX_mat, 
                            min.prop = 1, 
                            group = paste0(pb_GEX_mat@colData$diet,
                                           pb_GEX_mat@colData$crypt_type) %>%
                              as.factor(), 
                            min.count = 1, 
                            min.total.count = 6)

pb_GEX_mat <- pb_GEX_mat[keep,]
print("dimension of matrix genes/samples")
dim(pb_GEX_mat)
```

Do tmm normalization for regular exploratory analysis

```{r}
scale_factor = 100000
all_nf <- edgeR::calcNormFactors(pb_GEX_mat, method = "TMM")
sfs <- all_nf$samples$lib.size * all_nf$samples$norm.factors
pb <- sweep(assay(pb_GEX_mat, "counts"), MARGIN = 2, sfs, FUN = "/")
assay(pb_GEX_mat, "logcounts") <- log1p(pb * scale_factor)
```

Clearly there are many lowly expressed genes in most of the samples (flexible cut)

```{r}
boxplot(assay(pb_GEX_mat, "logcounts"))
```

Making data available:

```{r}
column_names <- colData(pb_GEX_mat)[,"id"]
pbulk_dat <- assay(pb_GEX_mat, "logcounts")
colnames(pbulk_dat) <- column_names
pbulk_dat %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  write_csv("./pub_results/pbulk_TMM.csv")
```

```{r}
column_names <- colData(pb_GEX_mat)[,"id"]
pbulk_dat <- assay(pb_GEX_mat, "counts")
colnames(pbulk_dat) <- column_names
pbulk_dat %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  write_csv("./pub_results/pbulk_counts.csv")
```

## Principal Component analysis

```{r}
PCA_gex <- prcomp(pb_GEX_mat %>% assay(., "logcounts") %>% t(), scale. = FALSE) %>% 
  summary()

# Principal components
rownames(PCA_gex$x) <- colData(pb_GEX_mat)$id 
```

## We are gonna cluster the pseudobulk samples based on all PCs 

The reason of not doing it with the first two components is because the amount of explained variance is still very little (~ 22%)

```{r}
sample_clust <- dist(PCA_gex$x) %>% hclust()
k2 <- cutree(sample_clust, 2) %>%
  enframe(name = "id", value = "hclust_2")
k4 <- cutree(sample_clust, 4) %>%
  enframe(name = "id", value = "hclust_4")
clust_info <- left_join(k2, k4, by ="id") %>%
  dplyr::mutate(hclust_2 = as.character(hclust_2),
                hclust_4 = as.character(hclust_4))
```

It seems that the diet has an interaction term with the granuloma crypt, nevertheless it may be the case that this separation has only something to do with the actual profiled genes or counts

```{r,fig.height=4, fig.width= 6.5, message = F, warning = FALSE}
pca_dat <- PCA_gex$x %>%
  as.data.frame() %>%
  rownames_to_column("id") %>%
  dplyr::select(id, PC1, PC2) %>%
  left_join(coldata, by = "id")

pca_plot <- pca_dat %>%
  ggplot(aes(x = PC1, y = PC2, color = diet)) +
  geom_point(size = 3, aes(shape = crypt_type)) +
  theme_classic() +
  theme(axis.text = element_text(size = 12)) +
  xlab(paste0("Prop. Var PC1 = ", PCA_gex$importance[2,1] %>% round(2))) +
  ylab(paste0("Prop. Var PC2 = ", PCA_gex$importance[2,2] %>% round(2)))

plot(pca_plot)
```

```{r}
pdf("./pub_results/PCA_plt.pdf", height = 3, width = 4.5)
plot(pca_plot)
dev.off()

write_csv(pca_dat,"./pub_results/PCA_plt.csv")
```

## What are the genes in the loadings of PC1 and PC2

```{r}
loadings <- PCA_gex$rotation[,c("PC1", "PC2")] %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, names_to = "PC") 

pos_loadings <- loadings %>%
  dplyr::filter(value > 0) %>%
  arrange(desc(value)) %>%
  group_by(PC) %>%
  dplyr::slice(1:25)

neg_loadings <- loadings %>%
  dplyr::filter(value < 0) %>%
  arrange(value) %>%
  group_by(PC) %>%
  dplyr::slice(1:25)

top_loadings <- bind_rows(pos_loadings, neg_loadings)

write_csv(loadings, "./pub_results/loadings_PCA.csv")
```

### PC1

These are the top 25 and bottom 25 genes in the loading matrix

```{r}
top_loadings %>%
  dplyr::filter(PC == "PC1") %>%
  arrange(value) %>%
  paged_table()
```

```{r}
top_loadings %>%
  dplyr::filter(PC == "PC2") %>%
  arrange(value) %>%
  paged_table()
```

Let's estimate the amount of variance explained by the crypt-type and diet alone

```{r}
pc_var <- PCA_gex$importance %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("PC") %>%
  dplyr::select(PC, `Proportion of Variance`)

colnames(pc_var) <- c("PC","Prop_var")
```

```{r}
vartest <- PCA_gex$x %>%
  as.data.frame() %>%
  rownames_to_column("id") %>%
  pivot_longer(-id, names_to = "PC") %>%
  left_join(coldata, by = "id") %>%
  group_by(PC) %>%
  nest() %>%
  mutate(diet_p = map(data, function(dat) {
    aov(value ~ diet,data = dat) %>%
      broom::tidy() %>%
      dplyr::filter(term == "diet") %>%
      dplyr::pull(p.value)
    
  })) %>%
  mutate(crypt_p = map(data, function(dat) {
    aov(value ~ crypt_type,data = dat) %>%
      broom::tidy() %>%
      dplyr::filter(term == "crypt_type") %>%
      dplyr::pull(p.value)
    
  })) %>%
  dplyr::select(PC, diet_p, crypt_p) %>%
  ungroup() %>%
  unnest() %>%
  left_join(pc_var)
```

### Proportion of explained variance per group

* Uncorrected p-values, just for reference

### Diet

```{r}
vartest %>%
  dplyr::filter(diet_p <= 0.05)  %>%
  pull(Prop_var) %>%
  sum()
```

### Crypt

```{r}
vartest %>%
  dplyr::filter(crypt_p <= 0.05)  %>%
  pull(Prop_var) %>%
  sum()
```

Both, the granuloma and the diet seem to influence the pseusobulk gene expression of crypts

## Running edgeR

First we will use only the top genes of both PC1 and PC2

```{r}
top_loadings <- loadings %>%
  dplyr::mutate(direction = ifelse(sign(value) > 0, 
                                   "pos", "neg")) %>%
  group_by(PC, direction) %>%
  mutate(cutoff_pos = quantile(abs(value), 0.70)) %>%
  dplyr::filter(abs(value) >= cutoff_pos)

useful_genes <- top_loadings %>% pull(gene) %>% unique()
```

### We will create multifactors

```{r}
pb_GEX_mat_old <- pb_GEX_mat
pb_GEX_mat <- pb_GEX_mat[useful_genes,]
```

```{r}
pb_GEX_mat_edge <- calcNormFactors(pb_GEX_mat)
meta <- pb_GEX_mat_edge$samples
meta$diet <- factor(meta$diet, levels = c("STD", "GW"))
meta$crypt_type <- factor(meta$crypt_type, levels = c("ctrl_crypt", "granuloma_crypt"))
meta$group <- paste0(meta$diet, ".", meta$crypt_type)

design <- model.matrix(~ 0 + group, meta)
pb_GEX_mat_edge <- estimateDisp(pb_GEX_mat_edge, design)
fit <- glmQLFit(pb_GEX_mat_edge, design, robust=TRUE)

saveRDS(fit, "./pub_results/edgeR_fit.rds")
```

```{r}
my_contrasts <- makeContrasts(
  granulomavcrpt = ((groupGW.granuloma_crypt + groupSTD.granuloma_crypt)/2) - ((groupGW.ctrl_crypt + groupSTD.ctrl_crypt)/2),
  diet_effect = ((groupGW.granuloma_crypt + groupGW.ctrl_crypt)/2) - ((groupSTD.granuloma_crypt + groupSTD.ctrl_crypt)/2),
  diet_effect_ctrl =  groupGW.ctrl_crypt - groupSTD.ctrl_crypt,
  dief_effect_granuloma = (groupGW.granuloma_crypt - groupGW.ctrl_crypt) - (groupSTD.granuloma_crypt - groupSTD.ctrl_crypt),
  diet_effect_granuloma_simple = groupGW.granuloma_crypt -groupSTD.granuloma_crypt, # Not considering change within control
  granuloma_effect_std = groupSTD.granuloma_crypt - groupSTD.ctrl_crypt,
  granuloma_effect_gw = groupGW.granuloma_crypt - groupGW.ctrl_crypt,
  levels=design)

saveRDS(my_contrasts, "./pub_results/edgeR_contrasts.rds")
```

### Effect of granuloma crypt

What are the up or downregulated genes in the granuloma vs the ctrl crypt? This is done across slides

```{r}
granuloma_degs <- glmQLFTest(fit, contrast=my_contrasts[,"granulomavcrpt"])

granuloma_degs <- granuloma_degs %>%
  topTags(., n = Inf) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  dplyr::filter(FDR <= 0.1)
```

A ranking was performed by a logFC * significance value and saved in a file

```{r}
granuloma_degs %>%
  dplyr::mutate(rank_stat = abs(logFC) * -log10(FDR)) %>%
  arrange(desc(rank_stat)) %>%
  dplyr::select(gene, rank_stat, logFC, FDR) %>%
  write_csv("./pub_results/granulomavcrpt_degs.csv")
```

```{r}
assay(pb_GEX_mat, "logcounts")["Defa30",] %>% enframe() %>%
  mutate(id = colData(pb_GEX_mat)$id) %>%
  left_join(coldata, by = "id") %>%
  ggplot(aes(x = crypt_type, y = value, color = diet)) +
  geom_boxplot() +
  theme_classic() +
  theme(axis.text = element_text(size = 12)) +
  ggtitle("Defa30")
```

The effects don't seem to be very clear, what if we focus solely in ?

### Effect of granuloma in crypts only in STD diet

```{r}
granuloma_degs <- glmQLFTest(fit, contrast=my_contrasts[,"granuloma_effect_std"])

granuloma_degs <- topTags(granuloma_degs, n = Inf) %>%
    as.data.frame() %>%
    rownames_to_column("gene") %>%
  dplyr::filter(FDR <= 0.1) %>%
  paged_table()
  
```

A ranking was performed by a logFC * significance value and saved in a file

```{r}
granuloma_degs %>%
  dplyr::mutate(rank_stat = abs(logFC) * -log10(FDR)) %>%
  arrange(desc(rank_stat)) %>%
  dplyr::select(gene, rank_stat, logFC, FDR) %>%
  write_csv("./pub_results/granuloma_effect_std_degs.csv")
```

### Effect of granuloma in crypts only in GW diet

```{r}
granuloma_degs <- glmQLFTest(fit, contrast=my_contrasts[,"granuloma_effect_gw"])

granuloma_degs <- topTags(granuloma_degs, n = Inf) %>%
    as.data.frame() %>%
    rownames_to_column("gene")
```

A ranking was performed by a logFC * significance value and saved in a file

```{r}
granuloma_degs %>%
  dplyr::mutate(rank_stat = abs(logFC) * -log10(FDR)) %>%
  arrange(desc(rank_stat)) %>%
  dplyr::select(gene, rank_stat, logFC, FDR) %>%
  dplyr::filter(FDR <= 0.1) %>%
  write_csv("./pub_results/granuloma_effect_gw_degs.csv")
```

### Effect of diet in all groups

What are the up or downregulated genes in the GW vs the STD diet?

```{r}
diet_degs <- glmQLFTest(fit, contrast=my_contrasts[,"diet_effect"])

diet_degs <- topTags(diet_degs, n = Inf) %>%
    as.data.frame() %>%
    rownames_to_column("gene")

diet_degs %>%
   dplyr::mutate(rank_stat = abs(logFC) * -log10(FDR)) %>%
  arrange(desc(rank_stat)) %>%
  dplyr::select(gene, rank_stat, logFC, FDR) %>%
  dplyr::filter(FDR <= 0.1) %>%
  write_csv("./pub_results/diet_effect_degs.csv")
```
